add support for new dependencyResolutionManagement

Logiquo opened the pull request.

Logiquo committed:
move repo declartions

- Move repository declartions in MavenConfiguration.java to LoomRepositoryPlugin.java

Logiquo committed:
move repo declartions

- Move repository declartions in MinecraftMappedProvider.java to LoomRepositoryPlugin.java

Logiquo committed:
move repo declartions

- Move repository declarations in MinecraftProcessedProvider.java to LoomRepositoryPlugin.java

Logiquo committed:
Merge branch 'FabricMC:dev/0.8' into dependencyResolutionManagement

Logiquo committed:
do not add repositories if dependencyResolutionManagement is used

Logiquo commented:
This is related to #395.

modmuss50 commented:
Had a quick look from my phone and looks good. 👍 can you add an integration test for this?

Let me know if you need help with that.

liach commented on src/main/java/net/fabricmc/loom/LoomGradlePlugin.java:
```diff
@@ -40,37 +41,43 @@
 	@Override
-	public void apply(Project project) {
-		project.getLogger().lifecycle("Fabric Loom: " + LoomGradlePlugin.class.getPackage().getImplementationVersion());
+	public void apply(PluginAware target) {
```
The diff can be simplified by retaining `apply(Project)` and adding a `apply(PluginAware)` at the same time. The override will move to the `apply(PluginAware)`, which will be:
```java
	@Override
	public void apply(PluginAware target) {
		target.getPlugins().apply(LoomRepositoryPlugin.class);
		if (target instanceof Project) {
			apply((Project) target);
		}
	}
```

liach commented on src/main/java/net/fabricmc/loom/configuration/processors/MinecraftProcessedProvider.java:
```diff
@@ -37,15 +39,20 @@
 
 	private final JarProcessorManager jarProcessorManager;
 
+	private static String getProjectUUID(Project project) {
```
Mind rename to `getProjectUuid` which is the fabric style?

liach commented on src/main/java/net/fabricmc/loom/configuration/LoomDependencyManager.java:
```diff
@@ -193,7 +194,9 @@ private static void handleInstallerJson(JsonObject jsonObject, Project project)
+			// these repositories manually in the settings file.
+			if (jsonElement.getAsJsonObject().has("url") && !project.getGradle().getPlugins().hasPlugin(LoomRepositoryPlugin.class)) {
 				String url = jsonElement.getAsJsonObject().get("url").getAsString();
 				long count = project.getRepositories().stream().filter(artifactRepository -> artifactRepository instanceof MavenArtifactRepository)
```
So here, when the count is 0, I think the best approach is to leave a warning or fail the build (maybe add a command-line option to control that) when the desired maven doesn't exist. Imo we should fail by default, but a system property for warning can allow people to use proxy mavens.

Logiquo replied:
The `settings.getDependencyResolutionManagement().getRepositories()` and `project.getRepositories()` are different object in gradle. Since we do not add any repo into `project.getRepositories()` if `dependencyResolutionManagement` used, the `project.getRepositories()` will always be empty. Furthermore, there is currently no way of retrieving `settings` in project lifecycle, this means we do not know if a repo is added or not. Thus, the `count` will always be zero. 

This limitation will cause your idea hard to implement, and hence I prefer the existing code (I really don't have any other solution for this).

Juuxel commented on src/main/java/net/fabricmc/loom/LoomGradlePlugin.java:
```diff
@@ -40,37 +41,43 @@
+		target.getPlugins().apply(LoomRepositoryPlugin.class);
 
-		refreshDeps = project.getGradle().getStartParameter().isRefreshDependencies();
+		if (target instanceof Project) {
+			Project project = (Project) target;
```
```suggestion
		if (target instanceof Project project) {
```

Juuxel commented on src/main/java/net/fabricmc/loom/configuration/processors/MinecraftProcessedProvider.java:
```diff
@@ -37,15 +39,20 @@
 
 public class MinecraftProcessedProvider extends MinecraftMappedProvider {
-	public static final String PROJECT_MAPPED_CLASSIFIER = "projectmapped";
+	public final String PROJECT_MAPPED_CLASSIFIER;
```
This should also follow the naming conventions (`projectMappedClassifier`)

Juuxel commented on src/main/java/net/fabricmc/loom/configuration/MavenConfiguration.java:
```diff
@@ -26,32 +26,7 @@
-import net.fabricmc.loom.LoomGradleExtension;
-
 public class MavenConfiguration {
 	public static void setup(Project project) {
```
I think this can just be removed now - all it did was set up those repos.

Juuxel commented on src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java:
```diff
@@ -0,0 +1,204 @@
+
+			// leave a marker so projects don't try to override these
+			settings.getGradle().getPluginManager().apply(LoomRepositoryPlugin.class);
+		} else if (target instanceof Project) {
+			Project project = (Project) target;
```
```suggestion
		} else if (target instanceof Project project) {
```

Juuxel commented on src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java:
```diff
@@ -0,0 +1,204 @@
+	public void apply(PluginAware target) {
+		RepositoryHandler repositories = null;
+
+		if (target instanceof Settings) {
+			Settings settings = (Settings) target;
```
```suggestion
		if (target instanceof Settings settings) {
```

Juuxel commented on src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java:
```diff
@@ -0,0 +1,204 @@
+	File getUserCache() {
+		File gradleUserHomeDir = null;
+
+		if (target instanceof Settings) {
+			gradleUserHomeDir = ((Settings) target).getGradle().getGradleUserHomeDir();
+		} else if (target instanceof Project) {
+			gradleUserHomeDir = ((Project) target).getGradle().getGradleUserHomeDir();
```
```suggestion
		if (target instanceof Settings settings) {
			gradleUserHomeDir = settings.getGradle().getGradleUserHomeDir();
		} else if (target instanceof Project project) {
			gradleUserHomeDir = project.getGradle().getGradleUserHomeDir();
```

Juuxel commented on src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java:
```diff
@@ -0,0 +1,204 @@
+	public File getRootPersistentCache() {
+		File rootDir = null;
+
+		if (target instanceof Settings) {
+			rootDir = ((Settings) target).getRootDir();
+		} else if (target instanceof Project) {
+			rootDir = ((Project) target).getRootDir();
```
```suggestion
		if (target instanceof Settings settings) {
			rootDir = settings.getRootDir();
		} else if (target instanceof Project project) {
			rootDir = project.getRootDir();
```

Juuxel commented on src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java:
```diff
@@ -0,0 +1,204 @@
+	public File getRootBuildCache() {
+		File rootDir = null;
+
+		if (target instanceof Settings) {
+			rootDir = ((Settings) target).getRootDir();
+		} else if (target instanceof Project) {
+			rootDir = ((Project) target).getRootDir();
```
```suggestion
		if (target instanceof Settings settings) {
			rootDir = settings.getRootDir();
		} else if (target instanceof Project project) {
			rootDir = project.getRootDir();
```

Logiquo commented:
> Had a quick look from my phone and looks good. 👍 can you add an integration test for this?
> 
> Let me know if you need help with that.

Hi, I did not have any experience in writing integration tests. If you can give me some examples/tutorials then I will be happy to write it.

Logiquo committed:
Simplify the change on LoomGradlePlugin

- this is the suggestion from liach

Logiquo committed:
change name to follow fabric naming convension

- change getProjectUUID to getProjectUuid
- change PROJECT_MAPPED_CLASSIFIER to projectMappedClassifier

Logiquo committed:
remove MavenConfiguration.java

- the file currently do nothing.

Logiquo committed:
clean-up for all `instanceof` clause

liach commented:
@LogicFan seems what you said in https://github.com/FabricMC/fabric-loom/issues/395#issuecomment-841046582 still needs to be addressed?

Logiquo commented:
> @LogicFan seems what you said in [#395 (comment)](https://github.com/FabricMC/fabric-loom/issues/395#issuecomment-841046582) still needs to be addressed?

I agree with your solution, however, I am still testing this part of the code to make sure the behaviour is expected. So a commit will be up for today or tomorrow after I am confident that my code is correct.

liach commented:
Also on unit tests, I think you will make a backbone project at https://github.com/FabricMC/fabric-loom/tree/dbc1f4cba7ad6c9bbce0aea46e682897a61c53dd/src/test/resources/projects and add a driver code file at https://github.com/FabricMC/fabric-loom/tree/dbc1f4cba7ad6c9bbce0aea46e682897a61c53dd/src/test/groovy/net/fabricmc/loom/test/integration

Logiquo committed:
add DependencyResolutionManagementTest

Logiquo commented:
The test is added. Waiting for approval.

Logiquo committed:
code cleanup

modmuss50 commented on src/test/resources/projects/dependencyResolutionManagement/projmap/src/main/resources/modid.accesswidener:
```diff
@@ -0,0 +1,2 @@
+accessWidener   v1   named
+extendable   class   FluidState
```
This isnt a valid accesswidener, I would remove it as I dont see how its realted to this test.

modmuss50 replied:
Oh, I see it forces a project mapped jar, still not a valid class name

Logiquo replied:
Sorry, I do not have much experience with how the accesswidener works. This test is to force a build to use a project-mapped jar. Could you change this to a valid accesswidener? Thanks.

modmuss50 commented on src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java:
```diff
@@ -0,0 +1,200 @@
+	}
+}
+
+final class Cache {
```
Might be out of scope for this PR, but it does duplicate some stuff in LoomGradleExtension, moving this out into its own class and removing the duplicates might not be a bad idea. 

At somepoint I want to give LGE a good cleanup so im not too fussed about doing this now.

modmuss50 commented on src/test/resources/projects/dependencyResolutionManagement/projmap/src/main/resources/modid.accesswidener:
```diff
@@ -0,0 +1,2 @@
+accessWidener   v1   named
+extendable   class   FluidState
```
```suggestion
extendable   class   net/minecraft/fluid/FluidState
```

modmuss50 committed:
Update src/test/resources/projects/dependencyResolutionManagement/projmap/src/main/resources/modid.accesswidener

Logiquo commented:
Hi, may I ask what is the ETA to merge this PR? Is there anything more I need to do? This is my first PR in GitHub so I do not have much experience.

modmuss50 commented:
Should be ready to merge, I just want to test it locally first.

modmuss50 commented on src/main/java/net/fabricmc/loom/configuration/processors/MinecraftProcessedProvider.java:
```diff
@@ -37,15 +39,20 @@
 	private final JarProcessorManager jarProcessorManager;
 
+	private static String getProjectUuid(Project project) {
+		return UUID.nameUUIDFromBytes(project.getProjectDir().toString().getBytes(StandardCharsets.UTF_8)).toString();
```
Im not sure if a whole UUID is needed? would just the project name (formatted in a classifier safe format) be a better solution?

Logiquo replied:
In theory, anything that unique for the project should be fine. UUID is just for naming collision avoidance. I don't know if the project name would be safe enough for this. Can people define two sub-projects with the same name?

liach replied:
I think if you use full subproject names, you are fine. You can define like `:sub1:duplicate` and `:sub2:duplicate` subprojects within one project I think, but their full project names differ.

Logiquo replied:
I see, I will change it to the fully-qualified-name. It will be more human-readable, and hopefully shorter. A commit will be up for this tomorrow.

Logiquo replied:
A commit is up. My biggest concern is what if there is invalid char (i.e. cannot exists in the path) inside the project name. Any solution for this? Should we write a transformer that replaces all the invalid char?

I cannot find a good resource for what is a valid name for the Gradle project, seems and char except `:` are legal.

Logiquo committed:
change project uuid to project full name
